### PR TITLE
Optimize memory usage of LayerNorm

### PR DESCRIPTION
**Pull request by daemyung**
## 🙏 Describe the pull request

- 메모리 사용량을 줄였습니다.
- 중복된 테스트를 통합하였습니다.

## ✅ Checklist

- [x] Code follows the project's coding conventions and style.
- [x] Tests have been added or updated to cover the changes.
- [x] Documentation has been updated, if necessary.

### CODE REVIEW

File: trident/operation/softmax.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ def grid(meta):
             return [num_vec]
 
         out = torch.empty_like(inp)
-        blk_sz = util.get_proper_block_size(vec_sz, inp.element_size())
+        blk_sz = util.get_block_size(vec_sz, inp.element_size())
         num_warps = math.clamp(triton.next_power_of_2(vec_sz // 512), 2, 32)
```
Comment (kakao-steve-ai) [inline trident/operation/softmax.py]: 여기는 `util.get_num_warps` 대신에 이렇게 계산하는 이유가 있나요?

File: trident/operation/instance_norm.py [hunk 2]
```diff
@@ -42,7 +42,7 @@ def grid(meta):
             return [num_batches * num_ch]
 
         out = torch.empty_like(inp)
-        vec_blk_sz = util.get_proper_block_size(vec_sz, inp.element_size())
+        vec_blk_sz = util.get_block_size(vec_sz, inp.element_size())
 
         kernel.InstanceNorm.forward[grid](inp, num_ch, vec_sz, eps, out, vec_blk_sz, util.map_dtype(dtype),
                                           num_warps=16)
```
Comment (kakao-steve-ai) [inline trident/operation/instance_norm.py]: 여기는 `util.get_num_warps` 대신에 16을 사용하는 이유가 있나요?

**Review by kakao-steve-ai — APPROVED** [review]
`mean`, `std`를 메모리에 저장하지 않고 backward에 다시 계산하는거 확인했습니다. 계산량은 늘어나도 메모리 접근이 줄어들어서 좋아보입니다. BatchNorm도 이런식으로 수정해야겠네요.

**Comment by daemyung** [comment]
> `mean`, `std`를 메모리에 저장하지 않고 backward에 다시 계산하는거 확인했습니다. 계산량은 늘어나도 메모리 접근이 줄어들어서 좋아보입니다. BatchNorm도 이런식으로 수정해야겠네요.

저도 이번 변경이 맞는 방향인지 고민했었습니다. 고민한 결과 저희는 훈련보다는 추론에 집중 해야한다고 생각했습니다. 그래서 구현시 추론에 최적화된 방향으로 구현하려고 합니다.

File: trident/operation/softmax.py (hunk 1 above)
Comment (daemyung) [reply trident/operation/softmax.py]: @ansteve 없습니다. 변경할께요.

File: trident/operation/instance_norm.py (hunk 2 above)
Comment (daemyung) [reply trident/operation/instance_norm.py]: 없습니다. 업데이트할께요.

File: trident/util/util.py [hunk 3]
```diff
@@ -29,6 +31,10 @@ def get_shared_memory_size_per_block():
     return 64 * 1024
 
 
-def get_proper_block_size(num_elem, elem_sz):
+def get_block_size(num_elem, elem_sz):
     return min(triton.next_power_of_2(num_elem), get_shared_memory_size_per_block() // elem_sz)
 
+
+def get_num_warps(num_elem, elem_sz):
```
Comment (mejai1206) [inline trident/util/util.py]: 이러한 수식이 나오게 된 배경이 궁금합니다.

Comment (daemyung) [reply trident/util/util.py]: 매우 실험적인 코드입니다. 수식이 나온 배경은 매우 단순합니다. `block`의 크기 만큼 `shared memory`가 사용됩니다. 예를 들어서 `block`이 매우 작으면 사용되지 않는 `shared memory`가 많겠죠? 그러므로 `shared memory`를 잘 활용하기 위해서 남는 만큼 스레드의 수를 증가시켜줍니다. 이 생각을 코드로 작성한거에요.
